## Урок 2

### Практическое задание:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Для работы с матрицами
from scipy.sparse import csr_matrix, coo_matrix

# Детерминированные алгоритмы
from implicit.nearest_neighbours import ItemItemRecommender, CosineRecommender, TFIDFRecommender, BM25Recommender

# Метрики
from implicit.evaluation import train_test_split
from implicit.evaluation import precision_at_k, mean_average_precision_at_k, AUC_at_k, ndcg_at_k

In [2]:
data = pd.read_csv('../data/retail_train.csv')
data.head(2)

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0         -0.6        1631        1          0.0                0.0  
1          0.0        1631        1          0.0                0.0

In [3]:
result = pd.read_csv('../data/preds.csv')
result.head(2)

user_id                                             actual  \
0        1  [821867, 834484, 856942, 865456, 889248, 90795...   
1        3  [835476, 851057, 872021, 878302, 879948, 90963...   

                            random_recommendation  \
0  [825803, 13189883, 7159184, 14074112, 2240328]   
1     [2026874, 13133877, 851779, 926063, 931777]   

                          popular_recommendation  \
0  [6534178, 6533889, 1029743, 6534166, 1082185]   
1  [6534178, 6533889, 1029743, 6534166, 1082185]   

                                     itemitem  \
0  [999999, 1082185, 981760, 1127831, 995242]   
1  [999999, 1082185, 981760, 1098066, 995242]   

                                        cosine  \
0  [1082185, 999999, 981760, 1127831, 1098066]   
1   [1082185, 1098066, 981760, 999999, 826249]   

                                         tfidf  \
0  [1082185, 981760, 1127831, 999999, 1098066]   
1   [1082185, 981760, 1098066, 826249, 999999]   

                                  own_purchases  
0   [999999, 1082185, 1029743, 995785, 1004906]  
1  [999999, 1082185, 1098066, 6534178, 1127831]

In [4]:
test_size_weeks = 3

data_train = data[data['week_no'] < data['week_no'].max() - test_size_weeks]
data_test = data[data['week_no'] >= data['week_no'].max() - test_size_weeks]

In [78]:
def precision_at_k(recommended_list, bought_list, k):
    
    bought_list = np.array(bought_list)
    recommended_list = np.array(recommended_list)[:k]
    
    flags = np.isin(bought_list, recommended_list)
    precision = flags.sum() / len(recommended_list)
    
    
    return precision


### Задание 0. Товар 999999
На вебинаре мы использовали товар 999999 - что это за товар?
- *Все товары которые не вошли в топ-популярных.*

Зачем он нужен?
- *Чтобы не потерять пользователей котрые покупают не популярные товары.*

Используя этот товар мы смещаем качество рекомендаций.
В какую сторону?
- *Используя этот товар мы ухудшаем качество наших рекомендаций.* 

Можно ли удалить этот товар?
- *Удалять этот товар не стоит, иначе мы получим покупателей которые ничего не купили.*

Уберите этот товар (**внимание**: это можно сделать разными способами!) и сравните с качеством на семинаре.

### Задание 1. Weighted Random Recommendation

Напишите код для случайных рекоммендаций, в которых вероятность рекомендовать товар прямо пропорциональна логарифму продаж
- Можно сэмплировать товары случайно, но пропорционально какому-либо весу
- Например, прямопропорционально популярности. вес = log(sales_sum товара)
- Придумайте пример 3 весов, посчитайте weighted_random_recommendation для разных весов

In [5]:
result = data_test.groupby('user_id')['item_id'].unique().reset_index()
result.columns=['user_id', 'actual']
result['actual'] = result['actual'].apply(lambda x: list(x))
result.head(10)

user_id                                             actual
0        1  [821867, 834484, 856942, 865456, 889248, 90795...
1        3  [835476, 851057, 872021, 878302, 879948, 90963...
2        6  [920308, 926804, 946489, 1006718, 1017061, 107...
3        7  [840386, 889774, 898068, 909714, 929067, 95347...
4        8  [835098, 872137, 910439, 924610, 992977, 10412...
5        9  [864335, 990865, 1029743, 9297474, 10457112, 8...
6       13  [6534178, 1104146, 829197, 840361, 862070, 884...
7       14  [840601, 867293, 933067, 951590, 952408, 96569...
8       15  [910439, 1082185, 959076, 1023958, 1082310, 13...
9       16                       [1062973, 1082185, 13007710]

In [ ]:
perm = data_train.quantity / sum (data_train.quantity)
perm

In [12]:
def weighted_random_recommendation(items, n):
    """Случайные рекоммендации"""
    perm = data_train.quantity / sum (data_train.quantity)
    print (perm)
    items = np.array(items)
    recs = np.random.choice(items, size=n, replace=False)
    
    return recs.tolist()

### Задание 2. Расчет метрик
Рассчитайте Precision@5 для каждого алгоритма (с вебинара и weighted_random_recommendation) с помощью функции из вебинара 1. Какой алгоритм показывает лучшее качество? Почему?

In [79]:
for a in result.columns[1:]:
    result[a]=  result[a].map(lambda x: x[1:-1].split(', ')).apply(lambda x: list(map(int, x)))

In [80]:
result.apply(lambda x: precision_at_k(x['random_recommendation'], x['actual'],  5), axis=1).mean()

0.0008814887365328109

In [81]:
result.apply(lambda x: precision_at_k(x['popular_recommendation'], x['actual'],  5), axis=1).mean()

0.15523996082272082

In [82]:
result.apply(lambda x: precision_at_k(x['itemitem'], x['actual'],  5), axis=1).mean()

0.13692458374142857

In [83]:
result.apply(lambda x: precision_at_k(x['cosine'], x['actual'],  5), axis=1).mean()

0.13290891283055686

In [84]:
result.apply(lambda x: precision_at_k(x['tfidf'], x['actual'],  5), axis=1).mean()

0.1389813907933383

In [85]:
result.apply(lambda x: precision_at_k(x['own_purchases'], x['actual'],  5), axis=1).mean()

0.17969311132876015